docs/RELEASE.md

```diff
@@ -0,0 +1,126 @@
+
+### Steps
+
+1. Developer opens a Merge Request with changed crates against `master`.
+2. They bump all changed crates according to SemVer.
+3. They bump all crates that export any changed types in their *public API*.
+4. They also bump all crates that inherit logic changes from relying on one of the bumped crates. 
+
+## Mainline Release
+
+Cadence: every two weeks. Responsible: Release Team.
+
+This process aims to release the `release` branch as a *Mainline* release every two weeks. It should eventually be automated.
+
+### Steps
+
+1. Check if process [Clobbering](#clobbering) needs to happen and do so, if that is the case.
+2. Check out the latest commit of `release`.
+3. Verify all CI checks of that commit.
+4. Update the `CHANGELOG.md` version and date. <!-- Here we could aggregate the PrDocs into the CHANGELOG -->
+5. Open a Merge Request against `release` for visibility.
+6. Check if there were any changes since the last release and abort, if not.
+7. Run `cargo semver-checks` and `cargo public-api` again to ensure that there are no SemVer breaks.
+8. Internal QA from the release team can happen here.
+9.  Do a dry-run release to ensure that it *should* work.
+10. Merge it into `release`.
+11. Verify all CI checks.
+12. Comment that a *Mainline* release will happen from the merged commit hash.
+13. Release all changed crates to crates.io.
+14. Create a release on GitHub.
+
+## Nightly Release
+
+Cadence: every day at 00:00 UTC+1. Responsible: Release Team
+
+This process aims to release the `master` branch as a *Nightly* release. The process can start at 00:00 UTC+1 and should automatically do the following steps.
+
+1. Check out the latest commit of branch `master`.
+2. Verify all CI checks of that commit.
+3. Compare this commit to the latest `nightly*` tag and abort if there are no changes detected.
+4. Set the version of all crates to `major.0.0-nightlyYYMMDD` where `major` is the last released `major` version of that crate plus one.
+5. Tag this commit as `nightlyYYMMDD`.
+9. Do a dry-run release to ensure that it *should* work.
+7. Push this tag (the commit will not belong to any branch).
+8. Announce the intent to do a *Nightly* release from that tag in the RelEng chat.
+9. Release all crates that had changed since the last nightly release to crates.io.
+10. Create a release on GitHub.
+
+## Clobbering
+
+Cadence: every 6th release (~3 months). Responsible: Release Team
+
+This process aims to bring branch `release` in sync with the latest audited commit of `master`. It is not done via a Merge Request but rather by just copying files. It should be automated.
+
+The following script is provided to do the clobbering.
+
+```bash
+git fetch
+git checkout release
+git reset --hard origin/audited
+git push --force release
+```
```

Comment:
I'm not sure about this clobber process, I think it will cause us to lose commit history of old releases?

Instead, I suggest to make clobbering mean creating a new commit on the `release` branch to make the tracked files match **exactly** with what's at the `audited` tag, but keep the history of the `release` branch intact.


This could be achieved like this:

```bash
# Switch to release branch
git checkout release

# Delete all tracked files in the working directory
git ls-files -z | xargs -0 rm -f

# Find and delete any empty directories
find . -type d -empty -delete

# Checkout all files from master into the working directory
git checkout master -- .

# Stage, commit, and push the working directory which now matches master 1:1
git add .
git commit -m "some release message"
git push
```

This method is similar to what Gav proposed at https://hackmd.io/XD28Qh0nSzylw1v9ktALMg?both, but will make sure any deleted files from `master` are also deleted when we make the changes to release.

Reply:
>I'm not sure about this clobber process, I think it will cause us to lose commit history of old releases?

Yes. But we dont backport to them. Or is your concern that the code will not easily be accessible?  
Then we could tag it with the version.

Reply:
Yes my concern is the code from old releases can be completely removed. 

Using `push --force` and relying only on tags seems dangerous to me, if we forget to make a tag or accidentally delete/overwrite one we lose commit history on remote. 

With my approach we get all the same benefits (I think?) without the danger of accidentally losing history. 

I'm not sure what it has to do with backporting.

Reply:
> if we forget to make a tag or accidentally delete/overwrite one we lose commit history on remote.

Git is a decentralized version control system. Whoever would have pushed this, can just go back using `git reflog`.

Reply:
Whatever way we are using, we will need to extend the steps anyway as we will also need to bump the crate versions properly etc. Similar to what we do for the mainline release.

Reply:
> Git is a decentralized version control system. Whoever would have pushed this, can just go back using git reflog.

If I'm trying to debug an old release I really don't want to be DMing people asking for them to push their old code up for me.. It should just be in the remote? What's the downside to my approach or Gav's approach where we retain commit history on `stable`?

Reply:
So with your approach we still have the backport MRs into the `release` branch?  
Can we not just push a tag with the release commit? I imagine the `release` branch would look quite weird otherwise, since we have all the backport MRs and then these "clobber commits" that basically revert and reapply a lot of stuff.

Reply:
Yeah good point, did had thought about the missing history that much!

Reply:
> So with your approach we still have the backport MRs into the release branch?

What makes you think we need to backport? I think there's still some confusion. 

My suggestion is very simple; every major release to `stable`, a single new commit is added to make it match the `master` branch 1:1 (clobbering). This can be entirely automated and does not require any rebasing/merge conflicts. No existing commits are overridden.

Reply:
This sounds very much like my original (The Clobber) proposal.

Reply:
Yep @gavofyork, it's a slightly tweaked version of your original proposal. See https://github.com/paritytech/polkadot-sdk/pull/2490#discussion_r1414045936.

> This method is similar to what Gav proposed at https://hackmd.io/XD28Qh0nSzylw1v9ktALMg?both, but will make sure any deleted files from master are also deleted when we make the changes to release.

Reply:
Okay I will change it.